pkg/resource/deploy/step_executor.go

```diff
@@ -172,8 +177,31 @@ func (se *stepExecutor) ExecuteParallel(antichain antichain) completionToken {
 
 // ExecuteRegisterResourceOutputs services a RegisterResourceOutputsEvent synchronously on the calling goroutine.
 func (se *stepExecutor) ExecuteRegisterResourceOutputs(e RegisterResourceOutputsEvent) error {
-	// Look up the final state in the pending registration list.
+	return se.executeRegisterResourceOutputs(e, false /* errored */, false /* finalizingStackOutputs */)
+}
+
+func (se *stepExecutor) executeRegisterResourceOutputs(
+	e RegisterResourceOutputsEvent,
+	errored,
+	finalizingStackOutputs bool,
+) error {
 	urn := e.URN()
+
+	if finalizingStackOutputs {
+		contract.Assertf(urn.Type() == resource.RootStackType, "expected a stack resource urn, got %v", urn)
+	}
+
+	// If we're not finalizing and we've received an event for the stack's outputs, save the event for finalization
+	// later. We finalize stack outputs at the end of the deployment, so we can determine whether or not the
+	// deployment succeeded. If the deployment was successful, we use the new stack outputs. If there were errors,
+	// we keep the old stack outputs.
+	if !finalizingStackOutputs && urn.Type() == resource.RootStackType {
```

Comment:
Just a note that with Dans stacks as programs code incoming you need to be more careful checking for the root stack. You want to use QualifiedType() == RootStackType if you mean only the root stack, and Type() == RootStackType if you mean _any_ stack. Yes we should rename `RootStackType`.

Reply:
Switched to `QualifiedType() == RootStackType`